File: ServerPhp/src/dc/DC0042.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0042 Controller: Accounting doc. line
 */

require_once(PATH_N21LIB."/Http/Controller.php");

class DC0042 extends Controller {

  public function __construct($request, $response, $session) {
    parent::__construct($request, $response, $session);
    $this->_initFields();
  }

private function preQuery(&$params, &$where) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ACCDOC_ID like :ACCDOC_ID";
      $params["ACCDOC_ID"] = $_REQUEST["QRY_ACCDOC_ID"];
    if (!empty($_REQUEST["QRY_CLIENT_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CLIENT_ID like :CLIENT_ID";
      $params["CLIENT_ID"] = $_REQUEST["QRY_CLIENT_ID"];
    }
    if (!empty($_REQUEST["QRY_CREATEDBY"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CREATEDBY like :CREATEDBY";
      $params["CREATEDBY"] = $_REQUEST["QRY_CREATEDBY"];
    }
    if (!empty($_REQUEST["QRY_CREATEDON"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CREATEDON like :CREATEDON";
      $params["CREATEDON"] = $_REQUEST["QRY_CREATEDON"];
    }
    if (!empty($_REQUEST["QRY_CR_ACCT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CR_ACCT like :CR_ACCT";
      $params["CR_ACCT"] = $_REQUEST["QRY_CR_ACCT"];
    }
    if (!empty($_REQUEST["QRY_CR_AMOUNT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CR_AMOUNT like :CR_AMOUNT";
      $params["CR_AMOUNT"] = $_REQUEST["QRY_CR_AMOUNT"];
    }
    if (!empty($_REQUEST["QRY_CURRENCY"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.CURRENCY like :CURRENCY";
      $params["CURRENCY"] = $_REQUEST["QRY_CURRENCY"];
    }
    if (!empty($_REQUEST["QRY_DB_ACCT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.DB_ACCT like :DB_ACCT";
      $params["DB_ACCT"] = $_REQUEST["QRY_DB_ACCT"];
    }
    if (!empty($_REQUEST["QRY_DB_AMOUNT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.DB_AMOUNT like :DB_AMOUNT";
      $params["DB_AMOUNT"] = $_REQUEST["QRY_DB_AMOUNT"];
    }
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_IS_GENERATED"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.IS_GENERATED like :IS_GENERATED";
      $params["IS_GENERATED"] = $_REQUEST["QRY_IS_GENERATED"];
    }
    if (!empty($_REQUEST["QRY_MODIFIEDBY"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.MODIFIEDBY like :MODIFIEDBY";
      $params["MODIFIEDBY"] = $_REQUEST["QRY_MODIFIEDBY"];
    }
    if (!empty($_REQUEST["QRY_MODIFIEDON"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.MODIFIEDON like :MODIFIEDON";
      $params["MODIFIEDON"] = $_REQUEST["QRY_MODIFIEDON"];
    }
    if (!empty($_REQUEST["QRY_NOTES"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.NOTES like :NOTES";
      $params["NOTES"] = $_REQUEST["QRY_NOTES"];
    }
    if (!empty($_REQUEST["QRY_ORIG_AMOUNT"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ORIG_AMOUNT like :ORIG_AMOUNT";
      $params["ORIG_AMOUNT"] = $_REQUEST["QRY_ORIG_AMOUNT"];
    }
    if (!empty($_REQUEST["QRY_ORIG_CURRENCY"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ORIG_CURRENCY like :ORIG_CURRENCY";
      $params["ORIG_CURRENCY"] = $_REQUEST["QRY_ORIG_CURRENCY"];
    }
    if (!empty($_REQUEST["QRY_XRATE"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.XRATE like :XRATE";
      $params["XRATE"] = $_REQUEST["QRY_XRATE"];
    }
}

public function doQuery() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),50);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ACCDOC_ID
                ,pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME
                ,t.CLIENT_ID
                ,t.CREATEDBY
                ,t.CREATEDON
                ,t.CR_ACCT
                ,t.CR_AMOUNT
                ,t.CURRENCY
                ,t.DB_ACCT
                ,t.DB_AMOUNT
                ,t.ID
                ,t.IS_GENERATED
                ,t.MODIFIEDBY
                ,t.MODIFIEDON
                ,t.NOTES
                ,t.ORIG_AMOUNT
                ,t.ORIG_CURRENCY
                ,t.XRATE
            from AC_ACCDOC_LINE t $where $orderByClause ";
    $this->logger->debug($sql);
    $rs = $this->getDbConn()->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $dataOut = $this->serializeCursor($rs,array_keys($this->fields), $this->getDataFormatFetch());
    if ($this->getDataFormatFetch() == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
} /* end function doQuery  */


public function doExport() {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),50);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ID
                ,t.ACCDOC_ID
                ,pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME
                ,t.CLIENT_ID
                ,t.NOTES
                ,t.DB_ACCT
                ,t.CR_ACCT
                ,t.DB_AMOUNT
                ,t.CR_AMOUNT
                ,t.CURRENCY
                ,t.XRATE
                ,t.CREATEDON
                ,t.CREATEDBY
                ,t.MODIFIEDON
                ,t.MODIFIEDBY
                ,t.IS_GENERATED
                ,t.ORIG_AMOUNT
                ,t.ORIG_CURRENCY
            from AC_ACCDOC_LINE t $where $orderByClause ";
    $rs = $this->getDbConn()->Execute($sql, $params);
    $rsCount = $this->getDbConn()->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    } else {
      $columns = array_keys($this->fields);
    }
    if ($this->getDataFormatExport() == "csv" ) {
      $dataOut = $this->serializeCursor($rs,$columns,"csv");
    } else {
      $dataOut = $this->serializeCursor($rs,$columns,"xml");
      $dataOut = "<records>".$dataOut."</records>";
      $dataOut = "<queryParams>".$this->serializeMap($params,"xml")."</queryParams>".$dataOut;
      $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fields,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fields,false)."</columnDef>".$dataOut;
      $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
      $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
      $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    }
    $this->beginExport();
    print $dataOut;
    $this->endExport();
} /* end function doExport  */

public function fetchRecord() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0042.fetchRecord().");}
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function fetchRecord */

public function doInsert() {
    $this->populateRecordFromRequest(); 
    $sql = "insert into AC_ACCDOC_LINE(
                 ACCDOC_ID
                ,CLIENT_ID
                ,CREATEDBY
                ,CREATEDON
                ,CR_ACCT
                ,CR_AMOUNT
                ,CURRENCY
                ,DB_ACCT
                ,DB_AMOUNT
                ,ID
                ,IS_GENERATED
                ,MODIFIEDBY
                ,NOTES
                ,ORIG_AMOUNT
                ,ORIG_CURRENCY
                ,XRATE
            ) values ( 
                 :ACCDOC_ID
                ,:CLIENT_ID
                ,:CREATEDBY
                ,:CREATEDON
                ,:CR_ACCT
                ,:CR_AMOUNT
                ,:CURRENCY
                ,:DB_ACCT
                ,:DB_AMOUNT
                ,:ID
                ,:IS_GENERATED
                ,:MODIFIEDBY
                ,:NOTES
                ,:ORIG_AMOUNT
                ,:ORIG_CURRENCY
                ,:XRATE
    )";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doInsert */

public function doUpdate() {
    $this->populateRecordFromRequest();
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0042.doUpdate().");}
    $sql = "update AC_ACCDOC_LINE set 
                 CR_AMOUNT=:CR_AMOUNT
                ,ID=:ID
                ,ACCDOC_ID=:ACCDOC_ID
                ,CLIENT_ID=:CLIENT_ID
                ,DB_ACCT=:DB_ACCT
                ,CR_ACCT=:CR_ACCT
                ,DB_AMOUNT=:DB_AMOUNT
                ,CURRENCY=:CURRENCY
                ,XRATE=:XRATE
                ,NOTES=:NOTES
    where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    $pkArray = array("ID" => $this->record["ID"],"ID" => $this->record["ID"]);
    $this->findByPk($pkArray, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doUpdate */

public function doDelete() {
    $this->record["ID"] = $this->getRequestParam("ID");
    if (empty($this->record["ID"])) { throw new Exception("Missing value for primary key field ID in DC0042.doDelete().");}
    $sql = "delete from AC_ACCDOC_LINE where 
           ID= :ID
    ";
    $stmt = $this->getDbConn()->prepare($sql);
    $this->getDbConn()->Execute($stmt, $this->record);
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function doDelete */

public function initNewRecord() {
    $this->populateRecordFromRequest();
    $this->record["_p_record_status"] = "insert";
    $this->setFieldInitialValues($this->record, "DC0042");
    print "{success:true, data:".json_encode($this->record)."}";
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                t.ACCDOC_ID
                ,pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME
                ,t.CLIENT_ID
                ,t.CREATEDBY
                ,t.CREATEDON
                ,t.CR_ACCT
                ,t.CR_AMOUNT
                ,t.CURRENCY
                ,t.DB_ACCT
                ,t.DB_AMOUNT
                ,t.ID
                ,t.IS_GENERATED
                ,t.MODIFIEDBY
                ,t.MODIFIEDON
                ,t.NOTES
                ,t.ORIG_AMOUNT
                ,t.ORIG_CURRENCY
                ,t.XRATE
            from AC_ACCDOC_LINE t
         where 
           t.ID= :ID
            ";
    $rs = $this->getDbConn()->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    CollectionUtils::copyMapValues($record, $row);
} /* end function findByPk  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
    $this->populateRecordFromRequest();
    print "{success:true, data:".json_encode($this->record)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
} /* end function callProcedure */


private function _initFields() {
  $this->fields = array(
  "ACCDOC_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"ACCDOC_NAME" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CLIENT_ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"CREATEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CREATEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"CR_ACCT" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"CR_AMOUNT" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"CURRENCY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"DB_ACCT" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"DB_AMOUNT" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"ID" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"IS_GENERATED" => array(parent::FLDPROP_DATA_TYPE => "BOOLEAN")
  ,"MODIFIEDBY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"MODIFIEDON" => array(parent::FLDPROP_DATA_TYPE => "DATE")
  ,"NOTES" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"ORIG_AMOUNT" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
  ,"ORIG_CURRENCY" => array(parent::FLDPROP_DATA_TYPE => "STRING")
  ,"XRATE" => array(parent::FLDPROP_DATA_TYPE => "NUMBER")
);
}

}
?>

File: trunk/ServerJava/src/net/nan21/ebs/dc/DC0042.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0042 DC Controller: Accounting doc. line
 */

package net.nan21.ebs.dc;


import java.sql.Connection;
import java.sql.ResultSet;
import java.util.Date;
import java.util.HashMap;
import java.util.Properties;
import javax.servlet.http.HttpServletResponse;
import net.nan21.ebs.lib.CollectionUtils;
import net.nan21.ebs.lib.AbstractDataControl;
import net.nan21.ebs.lib.FieldDef;
import net.nan21.ebs.lib.HttpRequest;
import net.nan21.ebs.lib.HttpSession;
import net.nan21.ebs.lib.IDataControl;
import net.nan21.ebs.lib.DbManager;

public class DC0042 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ACCDOC_ID like :ACCDOC_ID");
      this.queryParams.put("ACCDOC_ID",(String)this.request.getParam("QRY_ACCDOC_ID"));
    if (this.request.getParam("QRY_CLIENT_ID") != null && !this.request.getParam("QRY_CLIENT_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CLIENT_ID like :CLIENT_ID");
      this.queryParams.put("CLIENT_ID",(String)this.request.getParam("QRY_CLIENT_ID"));
    }
    if (this.request.getParam("QRY_CREATEDBY") != null && !this.request.getParam("QRY_CREATEDBY").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CREATEDBY like :CREATEDBY");
      this.queryParams.put("CREATEDBY",(String)this.request.getParam("QRY_CREATEDBY"));
    }
    if (this.request.getParam("QRY_CREATEDON") != null && !this.request.getParam("QRY_CREATEDON").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CREATEDON like :CREATEDON");
      this.queryParams.put("CREATEDON",(String)this.request.getParam("QRY_CREATEDON"));
    }
    if (this.request.getParam("QRY_CR_ACCT") != null && !this.request.getParam("QRY_CR_ACCT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CR_ACCT like :CR_ACCT");
      this.queryParams.put("CR_ACCT",(String)this.request.getParam("QRY_CR_ACCT"));
    }
    if (this.request.getParam("QRY_CR_AMOUNT") != null && !this.request.getParam("QRY_CR_AMOUNT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CR_AMOUNT like :CR_AMOUNT");
      this.queryParams.put("CR_AMOUNT",(String)this.request.getParam("QRY_CR_AMOUNT"));
    }
    if (this.request.getParam("QRY_CURRENCY") != null && !this.request.getParam("QRY_CURRENCY").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.CURRENCY like :CURRENCY");
      this.queryParams.put("CURRENCY",(String)this.request.getParam("QRY_CURRENCY"));
    }
    if (this.request.getParam("QRY_DB_ACCT") != null && !this.request.getParam("QRY_DB_ACCT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.DB_ACCT like :DB_ACCT");
      this.queryParams.put("DB_ACCT",(String)this.request.getParam("QRY_DB_ACCT"));
    }
    if (this.request.getParam("QRY_DB_AMOUNT") != null && !this.request.getParam("QRY_DB_AMOUNT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.DB_AMOUNT like :DB_AMOUNT");
      this.queryParams.put("DB_AMOUNT",(String)this.request.getParam("QRY_DB_AMOUNT"));
    }
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_IS_GENERATED") != null && !this.request.getParam("QRY_IS_GENERATED").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.IS_GENERATED like :IS_GENERATED");
      this.queryParams.put("IS_GENERATED",(String)this.request.getParam("QRY_IS_GENERATED"));
    }
    if (this.request.getParam("QRY_MODIFIEDBY") != null && !this.request.getParam("QRY_MODIFIEDBY").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.MODIFIEDBY like :MODIFIEDBY");
      this.queryParams.put("MODIFIEDBY",(String)this.request.getParam("QRY_MODIFIEDBY"));
    }
    if (this.request.getParam("QRY_MODIFIEDON") != null && !this.request.getParam("QRY_MODIFIEDON").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.MODIFIEDON like :MODIFIEDON");
      this.queryParams.put("MODIFIEDON",(String)this.request.getParam("QRY_MODIFIEDON"));
    }
    if (this.request.getParam("QRY_NOTES") != null && !this.request.getParam("QRY_NOTES").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.NOTES like :NOTES");
      this.queryParams.put("NOTES",(String)this.request.getParam("QRY_NOTES"));
    }
    if (this.request.getParam("QRY_ORIG_AMOUNT") != null && !this.request.getParam("QRY_ORIG_AMOUNT").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ORIG_AMOUNT like :ORIG_AMOUNT");
      this.queryParams.put("ORIG_AMOUNT",(String)this.request.getParam("QRY_ORIG_AMOUNT"));
    }
    if (this.request.getParam("QRY_ORIG_CURRENCY") != null && !this.request.getParam("QRY_ORIG_CURRENCY").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ORIG_CURRENCY like :ORIG_CURRENCY");
      this.queryParams.put("ORIG_CURRENCY",(String)this.request.getParam("QRY_ORIG_CURRENCY"));
    }
    if (this.request.getParam("QRY_XRATE") != null && !this.request.getParam("QRY_XRATE").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.XRATE like :XRATE");
      this.queryParams.put("XRATE",(String)this.request.getParam("QRY_XRATE"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.ACCDOC_ID"+
               " ,pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME"+
               " ,t.CLIENT_ID"+
               " ,t.CREATEDBY"+
               " ,to_char(t.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,t.CR_ACCT"+
               " ,t.CR_AMOUNT"+
               " ,t.CURRENCY"+
               " ,t.DB_ACCT"+
               " ,t.DB_AMOUNT"+
               " ,t.ID"+
               " ,t.IS_GENERATED"+
               " ,t.MODIFIEDBY"+
               " ,to_char(t.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,t.NOTES"+
               " ,t.ORIG_AMOUNT"+
               " ,t.ORIG_CURRENCY"+
               " ,t.XRATE"+
           " from AC_ACCDOC_LINE t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.ID"+
               " ,t.ACCDOC_ID"+
               ",pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME"+
               " ,t.CLIENT_ID"+
               " ,t.NOTES"+
               " ,t.DB_ACCT"+
               " ,t.CR_ACCT"+
               " ,t.DB_AMOUNT"+
               " ,t.CR_AMOUNT"+
               " ,t.CURRENCY"+
               " ,t.XRATE"+
               " ,to_char(t.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,t.CREATEDBY"+
               " ,to_char(t.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,t.MODIFIEDBY"+
               " ,t.IS_GENERATED"+
               " ,t.ORIG_AMOUNT"+
               " ,t.ORIG_CURRENCY"+
           " from AC_ACCDOC_LINE t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into AC_ACCDOC_LINE("+
               "  ACCDOC_ID"+
               " ,CLIENT_ID"+
               " ,CREATEDBY"+
               " ,CREATEDON"+
               " ,CR_ACCT"+
               " ,CR_AMOUNT"+
               " ,CURRENCY"+
               " ,DB_ACCT"+
               " ,DB_AMOUNT"+
               " ,ID"+
               " ,IS_GENERATED"+
               " ,MODIFIEDBY"+
               " ,NOTES"+
               " ,ORIG_AMOUNT"+
               " ,ORIG_CURRENCY"+
               " ,XRATE"+
           " ) values ( "+
               "  :ACCDOC_ID"+
               " ,:CLIENT_ID"+
               " ,:CREATEDBY"+
               " ,:CREATEDON"+
               " ,:CR_ACCT"+
               " ,:CR_AMOUNT"+
               " ,:CURRENCY"+
               " ,:DB_ACCT"+
               " ,:DB_AMOUNT"+
               " ,:ID"+
               " ,:IS_GENERATED"+
               " ,:MODIFIEDBY"+
               " ,:NOTES"+
               " ,:ORIG_AMOUNT"+
               " ,:ORIG_CURRENCY"+
               " ,:XRATE"+
    ")";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update AC_ACCDOC_LINE set "+
               "  CR_AMOUNT=:CR_AMOUNT"+
               " ,ID=:ID"+
               " ,ACCDOC_ID=:ACCDOC_ID"+
               " ,CLIENT_ID=:CLIENT_ID"+
               " ,DB_ACCT=:DB_ACCT"+
               " ,CR_ACCT=:CR_ACCT"+
               " ,DB_AMOUNT=:DB_AMOUNT"+
               " ,CURRENCY=:CURRENCY"+
               " ,XRATE=:XRATE"+
               " ,NOTES=:NOTES"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from AC_ACCDOC_LINE where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " t.ACCDOC_ID"+
                ",pbo_acc.get_accdoc_name_by_id(t.accdoc_id) ACCDOC_NAME"+
               " ,t.CLIENT_ID"+
               " ,t.CREATEDBY"+
               " ,to_char(t.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,t.CR_ACCT"+
               " ,t.CR_AMOUNT"+
               " ,t.CURRENCY"+
               " ,t.DB_ACCT"+
               " ,t.DB_AMOUNT"+
               " ,t.ID"+
               " ,t.IS_GENERATED"+
               " ,t.MODIFIEDBY"+
               " ,to_char(t.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,t.NOTES"+
               " ,t.ORIG_AMOUNT"+
               " ,t.ORIG_CURRENCY"+
               " ,t.XRATE"+
           " from AC_ACCDOC_LINE t"+
        " where "+
     "      t.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void callProcedure(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("ACCDOC_ID", new FieldDef("NUMBER"));
	  this.fields.put("ACCDOC_NAME", new FieldDef("STRING"));
	  this.fields.put("CLIENT_ID", new FieldDef("NUMBER"));
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.put("CR_ACCT", new FieldDef("STRING"));
	  this.fields.put("CR_AMOUNT", new FieldDef("NUMBER"));
	  this.fields.put("CURRENCY", new FieldDef("STRING"));
	  this.fields.put("DB_ACCT", new FieldDef("STRING"));
	  this.fields.put("DB_AMOUNT", new FieldDef("NUMBER"));
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("IS_GENERATED", new FieldDef("BOOLEAN"));
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("ORIG_AMOUNT", new FieldDef("NUMBER"));
	  this.fields.put("ORIG_CURRENCY", new FieldDef("STRING"));
	  this.fields.put("XRATE", new FieldDef("NUMBER"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	}

public void doCustomAction(String action) {}
}
